docs/configuration.md
# Configuration

When you run the server, the first (and only) argument should be the path to
the configuration file. This should be a [TOML](https://toml.io) file.

The following options are required:

- [`database.pass`](#user-and-pass)
- [`network.host`](#host)

You are also encouraged to check the following options:

- [`upload_dir`](#upload_dir)
- [`network.address`](#address)
- [`network.port`](#port)
- [`passwords`](#passwords)
- [`frontend_path`](#frontend_path)

You can find an example configuration file showcasing many of the options
[here](./config.example.toml).

## Top level options

These options go at the top level of the TOML file, not within any table:

### `frontend_path`

The path to a web frontend to use. This should point to a folder which contains
an `index.html` file and optionally a `static` folder.

The `index.html` file will be served at the root of the service. Files within
the `static` folder, if present, will be served under `/static`.

For example, if `frontend_path` was set to `/srv/shareit/frontend`, a file at
`/srv/shareit/frontend/static/img/logo.png` would be served at
`/static/img/logo.png`.

This setting may either be a string or `null` (the default). If it is `null`,
then no frontend will be served. Otherwise, it must be a path which the service
can read.

### `upload_dir`

The path to the directory where files will be uploaded. This should be a
directory that the service can read and write to. This defaults to
`/var/lib/shareit/shares`, but you may wish to change it to
`/srv/shareit/shares` or anything else.

This must be a string. It must be configured on Windows, but the default
may be suitable on Linux and other Unix-like systems.

### `highlighting_languages`

An array of languages that can be used for syntax highlighting of pastes.

The default value contains the 191 languages supported by
[highlight.js](https://highlightjs.org/) as of August 2021, as well as `auto`,
intended to indicate to the frontend that the language should be detected
automatically.

### `default_highlighting_language`

The default language for syntax highlighting of pastes, when one is not given.

The default for this is `auto`. It must be a string.

### `default_mime_type`

The default MIME type for uploaded files.

The default for this setting is `application/octet-stream`, which indicates
unknown binary data. It must be a string.

### `expiry_check_interval`

The interval at which the service will check for expired shares. This is only
used for clearing unused data (and share names), expired shares will still be
deleted if they are attempted to be accessed.

The default value for this is `1 min`. It must be a string. It supports a
variety of units, including `s[econds]`, `m[inutes]`, `h[ours]`, `d[ays]`,
`w[eeks]`, `M[onths]` and even `y[ears]`.

### `passwords`

A table mapping passwords to arrays of permissions.

The available permissions are:

- `create_link` - gives permission to create link shares.
- `create_file` - gives permission to create file shares.
- `create_paste` - gives permission to create paste shares.
- `create_any` - shorthand for `create_link`, `create_file`, and `create_paste`.
- `update_own` - allows people to update or delete their own shares.
- `update_any` - allows people to update or delete any share (implies `update_own`).
- `custom_name` - allows people to set a custom name for their shares.

Use the password `default` to describe permissions for users who do not
specify a password.

The default looks like:

```toml
[passwords]
default = ["create_any", "update_own", "custom_name"]
```

This allows anyone to create any share, using a custom name if they choose, and
to update or delete their own shares.

See [the example file](./config.example.toml) for a more complex example.

## Database options

These options go in a table named `database`. The database must be a PostgreSQL
database.

### `host` and `port`

The hostname (string) and port (integer) to access the database at. These
default to `localhost` and `5432` respectively.

### `user` and `pass`

The username and password to connect to the database with. These are strings.

The username defaults to `shareit`, but the password is required.

### `name`

This is the name of the database to use. It defaults to `shareit`, and must be
a string.

## Network options

These options go in a table named `network`.

### `address`

The address to bind to. This defaults to `127.0.0.1`, but you'll probably want
to change it to `0.0.0.0` to make it accessible from the Internet, unless you
are hosting the service behind a reverse proxy.

This must be a string.

### `port`

The port to listen on. This defaults to `8000`, but you may want to change it
to `80`, as this is the convential port for HTTP.

Note that SSL is not currently supported so port `443` is probably not
appropriate.

### `host`

The address at which this service can be accessed from the outside internet,
used for generating URLs to shares. This should include the scheme, which must
be either `http` or `https`.

This option is required, and must be a string.

## User restrictions

These options configure global limits on what users can do with the service.
They go in a table named `restrictions`.

### `max_upload_size`

The maximum file size for pastes and files. This defaults to `2 MB`. This must
be a string.

### `max_link_length`

The maximum length of a link to be shortened. This defaults to `255`, but can
be as high as `2047`. It must be an integer.

### `max_expiry_time`

The maximum time a share can be kept for. This can either be `null` (the
default), or a string.

If it is `null`, there is no limit, though a user can still specify a shorter
expiry time for their own shares. If it is a string, the same units are allowed
as for [`expiry_check_interval`](#expiry_check_interval).

### `allowed_mime_types` and `disallowed_mime_types`

These options allow you to restrict the MIME types for uploaded files.

They are both arrays of strings. If `allowed_mime_types` is not empty,
`disallowed_mime_types` will be ignored, and only the MIME types in
`allowed_mime_types` will be allowed.

Otherwise, all MIME types not in `disallowed_mime_types` will be allowed.

By default, `disallowed_mime_types` contains `text/html`, to prevent users
from uploading HTML files that look like web pages.

### `allowed_link_schemes`

The link schemes that are allowed for links to be shortened. This must be an
array of strings, and the default is `["http", "https"]`. To allow any scheme,
set it to the empty array (`[]`).

## Share naming options

These options configure how shares are named. They go in a table named `names`.

### `min_length` and `max_length`

The minimum and maximum length of a share name. These default to `1` and `32`
respectively. The maximum length that can be configured is `255`. These must
be integers.

### `random_length` and `random_attempt_limit`

These configure how random name generation behaves. The default for
`random_length` is `8` and the default for `random_attempt_limit` is `3`.

In short, `random_length` is the length to start generating random names with,
and `random_attempt_limit` is the number of random collisions that must occur
before increasing that length.

Here's a diagram to explain how random name generation works:

![Name generation diagram](./name_generation.png)

Or in text form:

1. Set the attempt count to 0.
2. Increase the attempt count by 1.
3. Generate a random name with the random name length.
4. If the name is available, use it, otherwise, continue.
5. If the attempt count is below the attempt limit, go back to step 2, otherwise, continue.
6. If increasing the default length is disabled, error, otherwise, continue.
7. If the random name length has reached the max length, error, otherwise, continue.
8. Increase the random name length by 1.
9. Go back to step 1.


src/config.rs
//! Manages configuration of the server and Rocket.
use byte_unit::Byte;
use rocket::config::{Environment, Limits};
use rocket::http::Status;
use rocket::response::status;
use serde::Deserialize;
use std::collections::HashMap;
use std::ffi::OsString;
use std::fs::{create_dir_all, File};
use std::io::Read;
use std::path::PathBuf;
use std::sync::RwLock;
use std::time::Duration;
use std::{env, process};

pub const DEFAULT_PASSWORD: &str = "default";

fn default_highlighting_language() -> String { "auto".into() }
fn default_mime_type() -> String { "application/octet-stream".into() }
fn default_expiry_check_interval() -> Duration { Duration::from_secs(60) }
fn default_min_name_length() -> u8 { 1 }
fn default_max_name_length() -> u8 { 32 }
fn default_random_name_length() -> RwLock<usize> { RwLock::new(8) }
fn default_random_name_attempt_limit() -> u8 { 3 }
fn default_max_upload_size() -> Byte { Byte::from_str("2 MB").unwrap() }
fn default_max_link_length() -> u16 { 255 }
fn default_allowed_link_schemes() -> Vec<String> { vec!["http".into(), "https".into()] }
fn default_disallowed_mime_types() -> Vec<String> { vec!["text/html".into()] }
fn default_bind_address() -> String { "127.0.0.1".into() }
fn default_bind_port() -> u16 { 8000 }
fn default_db_host() -> String { "127.0.0.1".into() }
fn default_db_port() -> u16 { 5432 }
fn default_db_user() -> String { "shareit".into() }
fn default_db_name() -> String { "shareit".into() }
fn default_upload_dir() -> PathBuf { "/var/shareit/shares/".into() }

fn default_passwords() -> HashMap<String, Vec<Permission>> {
    HashMap::from([(
        DEFAULT_PASSWORD.into(),
        vec![
            Permission::CreateAny,
            Permission::UpdateOwn,
            Permission::CustomName,
        ],
    )])
}

fn default_highlighting_languages() -> Vec<String> {
    include_str!("res/languages.txt")
        .split('\n')
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
        .collect()
}

#[derive(Deserialize)]
pub struct Config {
    #[serde(default)]
    pub frontend_path: Option<PathBuf>,
    #[serde(default = "default_upload_dir")]
    pub upload_dir: PathBuf,
    #[serde(default = "default_highlighting_languages")]
    pub highlighting_languages: Vec<String>,
    #[serde(default = "default_highlighting_language")]
    pub default_highlighting_language: String,
    #[serde(default = "default_mime_type")]
    pub default_mime_type: String,
    #[serde(with = "humantime_serde", default = "default_expiry_check_interval")]
    pub expiry_check_interval: Duration,
    #[serde(default = "default_passwords")]
    pub passwords: HashMap<String, Vec<Permission>>,
    #[serde(default)]
    pub names: NamesConfig,
    #[serde(default)]
    pub restrictions: RestrictionsConfig,
    pub network: NetworkConfig,
    pub database: DatabaseConfig,
}

#[derive(Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum Permission {
    CreateAny,
    CreateLink,
    CreateFile,
    CreatePaste,
    UpdateOwn,
    UpdateAny,
    CustomName,
}

#[derive(Deserialize)]
pub struct NamesConfig {
    #[serde(default = "default_min_name_length")]
    pub min_length: u8,
    #[serde(default = "default_max_name_length")]
    pub max_length: u8,
    #[serde(default = "default_random_name_length")]
    random_length: RwLock<usize>,
    #[serde(default = "default_random_name_attempt_limit")]
    pub random_attempt_limit: u8,
}

impl NamesConfig {
    pub fn get_random_length(&self) -> Result<usize, status::Custom<String>> {
        self.random_length
            .read()
            .map_err(|_| {
                status::Custom(Status::InternalServerError, "Internal server error.".into())
            })
            .map(|value| *value)
    }

    pub fn incr_random_length(&self) -> Result<(), status::Custom<String>> {
        let mut value = self.random_length.write().map_err(|_| {
            status::Custom(Status::InternalServerError, "Internal server error.".into())
        })?;
        *value += 1;
        Ok(())
    }
}

impl Default for NamesConfig {
    fn default() -> NamesConfig {
        NamesConfig {
            min_length: default_min_name_length(),
            max_length: default_max_name_length(),
            random_length: default_random_name_length(),
            random_attempt_limit: default_random_name_attempt_limit(),
        }
    }
}

#[derive(Deserialize)]
pub struct RestrictionsConfig {
    #[serde(default = "default_max_upload_size")]
    pub max_upload_size: Byte,
    // Note: Link length cannot be more than either max upload size or 2047.
    #[serde(default = "default_max_link_length")]
    pub max_link_length: u16,
    #[serde(with = "humantime_serde", default)]
    pub max_expiry_time: Option<Duration>,
    #[serde(default)]
    pub allowed_mime_types: Vec<String>,
    #[serde(default = "default_disallowed_mime_types")]
    pub disallowed_mime_types: Vec<String>,
    #[serde(default = "default_allowed_link_schemes")]
    pub allowed_link_schemes: Vec<String>,
}

impl Default for RestrictionsConfig {
    fn default() -> Self {
        RestrictionsConfig {
            max_upload_size: default_max_upload_size(),
            max_link_length: default_max_link_length(),
            max_expiry_time: None,
            allowed_mime_types: vec![],
            disallowed_mime_types: vec![],
            allowed_link_schemes: default_allowed_link_schemes(),
        }
    }
}

#[derive(Deserialize)]
pub struct NetworkConfig {
    pub host: url::Url,
    #[serde(default = "default_bind_address")]
    pub address: String,
    #[serde(default = "default_bind_port")]
    pub port: u16,
}

#[derive(Deserialize)]
pub struct DatabaseConfig {
    pub pass: String,
    #[serde(default = "default_db_host")]
    pub host: String,
    #[serde(default = "default_db_port")]
    pub port: u16,
    #[serde(default = "default_db_user")]
    pub user: String,
    #[serde(default = "default_db_name")]
    pub name: String,
}

impl Config {
    pub fn load() -> Self {
        let args: Vec<OsString> = env::args_os().collect();
        if args.len() < 2 {
            eprintln!("Usage: {} <config file>", args[0].to_string_lossy());
            process::exit(1);
        }
        let mut file =
            File::open(args[1].clone()).expect("Could not open config file (shareit.toml)");
        let mut raw = String::new();
        file.read_to_string(&mut raw)
            .expect("Could not read or decode config file");
        let config: Config = toml::from_str(&raw).expect("Could not parse config file");
        create_dir_all(&config.upload_dir).expect("Could not create upload directory");
        config
    }

    pub fn configure_rocket(&self) -> rocket::config::Config {
        let mut databases: HashMap<String, _> = HashMap::new();
        let mut database: HashMap<String, _> = HashMap::new();
        database.insert("url".into(), self.make_database_url());
        databases.insert("database".into(), database);
        rocket::config::Config::build(Environment::Production)
            .extra("databases", databases)
            .address(self.network.address.clone())
            .port(self.network.port)
            .limits(self.make_rocket_limits())
            .finalize()
            .expect("Could not configure Rocket")
    }

    pub fn make_database_url(&self) -> String {
        format!(
            "postgres://{}:{}@{}:{}/{}",
            self.database.user,
            self.database.pass,
            self.database.host,
            self.database.port,
            self.database.name
        )
    }

    fn make_rocket_limits(&self) -> Limits {
        let limit = self.restrictions.max_upload_size.get_bytes();
        Limits::new()
            .limit("form", limit)
            .limit("data-form", limit)
            .limit("file", limit)
            .limit("string", limit)
            .limit("bytes", limit)
    }
}
